### ISSUE TITLE
Add support for management agents with multiple partitions

### ISSUE THREAD

**Issue by Identitry** [lithnet/miis-autosync#102]
I'm thrilled over what you've accomplished, this is exactly what we've been looking for and been puzzeled MS hasn't supplied for us already. Thanks a lot!!!

Not been able to test it yet however reading the wiki I can't find a way of managing multi-partition MA's. For example in one of the solutions I'm working on we have a Generic SQL MA with three different object types involved that end's up in three different partitions/sets of run profiles and we also have a custom built MS Powershell MA that creates/deletes home folders on servers/sites in about 40 different locations around the world/partitions/sets of run profiles, that's a lot of run profiles! :-)

Maybe I've missed something but it would be great if an MA could be configured to have run profiles configured for each partition available! Maybe a bit to much to ask for now when the solution is still very young but to cover a complete solution this would be great. If this is something that already can be accomplished with the current release and I've missed it I would gladly recieve enlightment with maybe a wiki page regarding it.

**Comment by ryannewington** [comment]
Hmm. I've definitely built this with the assumption of a single-partition MA. I've got to get my head around partitions and how they work. The tool definitely needs to support, or at least be aware of multi-partitions. I'll need to spin up some multi-partition MAs in my environment to test. Do you have any recommendations on an easy to set up MA that supports partitions?

**Comment by Identitry** [comment]
I guess AD is the simplest one to start with, you could for example add the Configuration partition and start reading the CN=Sites container if you have multiple sites available otherwise maybe the CN=Partitions container.

**Comment by ryannewington** [comment]
I'll need something i can easily read and write to. I've got to generate various changes in the connector space to see how they are represented, in order to know how to respond to them. 

I added an FAQ question for now to say that multi-partitions are not supported yet, but I'll work on enabling this asap.

**Comment by Identitry** [comment]
Then I recommend the Generic SQL connector, not the easiest one to wrap your head around but a really good example since it uses a LDAP naming scheme for example CN=Henrik.Nilsson,Object=User and a second object could end up in CN=All Users,Object=Group.

If it would be any help, I'd love to help out and I'll supply you a simple SQL Database with the example object types above (two partitions) with internal mappings and a ready to use exported Generic SQL Connector, how about that? This would spare you the time to mess around with the Generic SQL Connector that could be hard unless you've done it before.

**Comment by ryannewington** [comment]
That would be an amazing help! I'd really appreciate that. That would certainly short-cut my discovery work.

**Comment by Identitry** [comment]
Great! I'll make sure to prepare it ASAP but maybe not today...

**Comment by marcusmaker** [comment]
Really great work! But as @Identitry says, support for multi-partition MA would be great. We use the new Generic SQL MA a lot, and for each table/object you are dealing with, it creates a new partition.

**Comment by ryannewington** [comment]
@Identitry @marcusmaker Are one of you able to provide me some run history XML for a multi-partition MA?

Specifically, I'm looking for run history that captures
1. at least two import operations from two different partitions on a multiple partition MA. 
2. A sync on an MA that results in staged exports on two different partitions of the same MA

You can email them through to me ryan@lithnet.io

Ryan

**Comment by weyto** [comment]
@ryannewington Send you the requested informations with demo data. I hope this can help you.

Thomas

**Comment by ryannewington** [comment]
Hi @weyto @Identitry @marcusmaker 

I've created the initial build with support for partitions. @Identitry this also includes a workaround for #104.

The UI elements arent polished yet, but should be functional. There are a few known issues at this stage (#105  #106 #107), but the new code should be functional enough for thorough testing.

So a few notes with this build
1. Missing partition support was a big oversight, and as such I had to re-do the way the internal controllers for each MA worked. I'll need to do a lot of regression testing with this build before release, as it changes the way even single-partition MAs worked

2. Detecting pending exports is tricky in a multi-partition environment. A sync on MA X that stages exports on MA Y doesn't give any clues in the run history as to what partitions were impacted on MA Y. So we don't automatically know what to export. I've implemented some logic to walk the connector space of pending exports to identify the partitions that require export (in as few steps as possible). I was initially just going to trigger an export on all partitions when a sync staged exports, but this wouldn't make sense for @Identitry 's scenario of 40 home folder locations throughout the world. In this build, I've added some trace logging and timers in to see how long this discovery actually takes in the real world. If it's too costly, I'll have to see if there is another way. 

Could I ask you guys to run it up in your environments, do some testing, and gather the autosync.log files and send them to me via email? If you have an active environment that will see regular changes, then leave it run for a couple of days? There's a lot of trace logging in there that will help me understand if the correct behaviours are taking place. Log any bugs as new issues here in github, and i'll attack them as they come in. 

@khandelwalpiyush we'll need to do a full regression test of the change detection mechanism from dev->prod in the mean time.

[autosync.zip](https://github.com/lithnet/miis-autosync/files/1346800/autosync.zip)

**Comment by ryannewington** [comment]
Oh, almost forgot.

This version will automatically update the schema of the config file when its loaded.

As the old version had run profiles stored against the MA, this version has them against the partitions, so the whole file format had to change.

I'd really appreciate a copy of the config.xml file (located in program files\lithnet\autosync) before and after you install this new version, so I can ensure the upgrade happened correctly (any creds will be encrypted and I wont be able to decrypt them).

Thanks!

**Comment by ryannewington** [comment]
Here is the build 
[autosync.zip](https://github.com/lithnet/miis-autosync/files/1346874/autosync.zip)

**Comment by weyto** [comment]
@ryannewington : Thx for your invest on the new Build. Is it possible that on the MA Controller only the last partition can be assigned to Run profiles?
I have build up multiple Import Run Profiles (for every partition one) put on the UI only the last partition can be assigned to the run profiles. The other partition only shows up with (none).
An other thing:
What is when i build up reference Attribut over multiple partitions? How could we ensure that after a import the referenced object it is also imported? Without autosync we can ensure that by setup one import profile with multiple steps. But this isn't supported jet by autosync (if i remember correctly).
Actual i want to test this use case, but based on the first explanation i can't do that.

Best regards,
Thomas

**Comment by ryannewington** [comment]
@weyto can you send me the autosync.log file so I can see what happened with the partition discovery. As long as you have created single-step profiles for each partition, they should appear.

References over multiple partitions is supported, but rather than creating a multi-step run profile, you just need to trigger imports on both partitions. If you always want to import from all partitions at the same time, you can use a PowerShell trigger to fire off the run profiles in order that you want to execute them.

When an event was detected, you can just output multiple run profile names at the end of your PowerShell trigger like this

```
write-output "DI (partition 1)"
write-output "DI (partition 2)"
```

**Comment by ryannewington** [comment]
@weyto you can actually use an repeating interval trigger to run a multi-step run profile. You could trigger your imports all in one go. AutoSync will actually work out which partitions need a sync after that.

It's really the export/confirming import/sync type profiles that need to be single-step only. AutoSync already knows what needs to be done, and wants to run a single profile. Running more than one step at a time here would just throw things into confusion.

**Comment by weyto** [comment]
@ryannewington thx for your additional infos. It was really my misunderstanding of the concept. Your really write with the idea of the PowerShell trigger.
The other thing with the profiles not showing up in the gui was a problem between me and the keyboard. (I don't explain more here :-)).
I have installed the version and do some tests. I send you the logs in the next days over mail.

**Comment by ryannewington** [comment]
Here is the most recent build with some bug fixes in it
[autosync.zip](https://github.com/lithnet/miis-autosync/files/1357778/autosync.zip)

**Comment by ryannewington** [comment]
@weyto @Identitry @marcusmaker any feedback on the new build?

**Comment by weyto** [comment]
@ryannewington I have some feedback about Reference values:
Use Case:
- 2 Partition P1 and P2 
- P1 have an Object-A and P2 have an Object-B
- Object-A have a reference to Object-B

In this case is the function "Added "Run Profile (Delta Sync)" to the front of the execution queue" not really the best way. In this case i can be that Object-B is not in the Metaverse and in this case the sync/export mechanism is triggering "to many times". If the Auto-Sync service have a feature for Multi-Partition Management Agents to disable the in front mechanism and will wait for complete import and sync of all Partition, then it will be only necessary to export one-time (to all other Management Agents).

What do you sync about that?

**Comment by ryannewington** [comment]
Hi @weyto 

Ok I think I understand the scenario, and can see why that would not be ideal. What is the mechanism you are using to trigger the import operations? I can certainly expose the 'move to front' setting, but it was added to make the execution order a bit more predictable. I'm wondering if we actually expose something on the trigger to tell it to defer the follow up operations.

Ryan

**Comment by weyto** [comment]
For all of my tests I use a simple PowerShell Trigger like this:

Sleep (60 * 1);    
Write-Output "FISO Company";
Write-Output "FISO OrganizationalUnits";	
Write-Output "FISO Employee";

I think your idea are really well to expose this functionality.
I'm really a fan of the mechanism, also of the 'move to front'. And for the most scenarios this works really well.
But in this special case multi partition MA with references over partition there are some optimizations.
All other Test cases passes from my site.
I can send you the logs if needed, but there are really many.

**Comment by ryannewington** [comment]
I know I said only use single-step run profiles, but perhaps try creating a run profile with the 3 import steps, and execute that using the PowerShell trigger. Some of the changes I've made to the latest version should be able to process a run of multiple import operations.

Can you send me the autosync.log file from the import event and subsequent actions?

**Comment by ryannewington** [comment]
Latest version based on feedback and bugs identified by @weyto 
[autosync.zip](https://github.com/lithnet/miis-autosync/files/1378529/autosync.zip)

**Comment by ryannewington** [comment]
Hi all. I've released public build [1.1.6500](https://github.com/lithnet/miis-autosync/releases/tag/v1.1.6500) with support for multiple partitions. Please let me know how you go with this build, and feel free to open any new issues if you run into problems.
